File: 백준/Silver IV/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys
a=list(sys.stdin.readline().strip('\n'))
while(a!=['.']):
  if(a[len(a)-1]!='.'):
    b=list(sys.stdin.readline().strip('\n'))
    a=a+b
  else:
    x=[]
    result=0
    for i in a:
      if(i=='(' or i=='['):
        x.append(i)
      elif(i==')'):
        if(len(x)>0 and x[len(x)-1]=='('):
          x.pop()
        else:
          result=1
          break
      elif(i==']'):
        if(len(x)>0 and x[len(x)-1]=='['):
          x.pop()
        else:
          result=1
          break
    if(result==0 and len(x)==0):
      print("yes")
    else:
      print("no")
    a=list(sys.stdin.readline().strip('\n'))

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        String p;
        Stack<Character> stack;
        boolean isit;

        while (true) {
            p = br.readLine();

            if (p.equals(".")) {
                break;
            }

            stack = new Stack<>();
            isit = true;

            for (int i = 0; i < p.length(); i++) {
                char ch = p.charAt(i);

                if (ch == '(' || ch == '[') {
                    stack.push(ch);
                } else if (ch == ')') {
                    if (stack.isEmpty() || stack.peek() != '(') {
                        isit = false;
                        break;
                    }
                    stack.pop();
                } else if (ch == ']') {
                    if (stack.isEmpty() || stack.peek() != '[') {
                        isit = false;
                        break;
                    }
                    stack.pop();
                }
            }

            if (!isit || !stack.isEmpty()) {
                bw.write("no\n");
            } else {
                bw.write("yes\n");
            }
        }

        bw.flush();
        bw.close();
        br.close();
    }
}